### PR TITLE
Focus a channel by joining it, refactor user commands #1189

### PR DESCRIPTION
**Pull request by realies**
This should work for focusing channels by joining them, the switch statement might need further refactoring. Thoughts?

### CODE REVIEW

File: client/js/lounge.js [hunk 1]
```diff
@@ -194,19 +195,27 @@ $(function() {
...
-		}
-
-		if (text.indexOf("/expand") === 0) {
-			$(".chan.active .toggle-button:not(.opened)").click();
-			return;
+		if (text.indexOf("/") === 0) {
+			const separatorPos = text.indexOf(" ");
+			const cmd = text.substring(1, separatorPos > 1 ? separatorPos : text.length);
+			const parameters = separatorPos > text.indexOf(cmd) ? text.substring(text.indexOf(cmd) + cmd.length + 1, text.length) : "";
+			switch (cmd) {
```
Comment (xPaw) [inline client/js/lounge.js]: This can be simplified like `if (cmd in modules)` or `if (typeof modules[cmd] === "function")`

File: client/js/modules.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
+"use strict";
```
Comment (YaManicKill) [inline client/js/modules.js]: Why is this stuff not just in utils?

File: client/js/lounge.js [hunk 3]
```diff
@@ -194,19 +195,27 @@ $(function() {
...
-			$(".chan.active .toggle-button.opened").click();
-			return;
-		}
-
-		if (text.indexOf("/expand") === 0) {
-			$(".chan.active .toggle-button:not(.opened)").click();
-			return;
+		if (text.indexOf("/") === 0) {
+			const separatorPos = text.indexOf(" ");
+			const cmd = text.substring(1, separatorPos > 1 ? separatorPos : text.length);
```
Comment (YaManicKill) [inline client/js/lounge.js]: `.split(" ")` is probably nicer.

File: client/js/modules.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+	utils.clear();
+}
+
+function collapse() {
+	$(".chan.active .toggle-button.opened").click();
+}
+
+function expand() {
+	$(".chan.active .toggle-button:not(.opened)").click();
+}
```
Comment (astorije) [inline client/js/modules.js]: These are used only once, I'd rather not have the extra level of indirection.

File: client/js/modules.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+module.exports = {
+	clear,
+	collapse,
+	expand,
+	join
+};
+
+function clear() {
+	utils.clear();
+}
```
Comment (astorije) [inline client/js/modules.js]: This seems like overkill 😅

File: client/js/modules.js (hunk 2 above)
Comment (astorije) [reply client/js/modules.js]: ~I agree.~ Actually, all those are just used once, I think this entire file/refactor is not wanted.

File: client/js/modules.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+	$(".chan.active .toggle-button:not(.opened)").click();
+}
+
+function join(channel) {
+	var chan = utils.findCurrentNetworkChan(channel);
+
+	if (chan.length) {
+		chan.click();
+		return true;
+	}
```
Comment (astorije) [inline client/js/modules.js]: Okay so _this_ is what this PR/commit is about :)
Whenever possible, try to separate refactors from actual changes in different commits, that makes it for a better history and easier review :)

**Comment by astorije** [comment]
@realies, are you willing/able to pursue work on this? :)

File: client/js/lounge.js (hunk 3 above)
Comment (realies) [reply client/js/lounge.js]: Please elaborate.

File: client/js/lounge.js [hunk 7]
```diff
@@ -184,14 +184,23 @@ $(function() {
...
-		}
-
-		if (text.indexOf("/expand") === 0) {
-			$(".chan.active .toggle-preview:not(.opened)").click();
-			return;
+		if (text.indexOf("/") === 0) {
+			const args = text.substr(1).split(" ");
+			const cmd = args.shift().toLowerCase();
+			if (typeof utils.inputCommands[cmd] === "function") {
+				if (cmd === "join") {
```
Comment (xPaw) [inline client/js/lounge.js]: Just pass the args into the command function, don't do the if check.

Join command can just use `args[0]`.

File: client/js/lounge.js [hunk 8]
```diff
@@ -184,14 +184,14 @@ $(function() {
...
-		}
-
-		if (text.indexOf("/expand") === 0) {
-			$(".chan.active .toggle-preview:not(.opened)").click();
-			return;
+		if (text.indexOf("/") === 0) {
+			const args = text.substr(1).split(" ");
+			const cmd = args.shift().toLowerCase();
+			if (typeof utils.inputCommands[cmd] === "function") {
+				if (utils.inputCommands[cmd](args)) {
```
Comment (xPaw) [inline client/js/lounge.js]: You can combine the ifs with `&&` to reduce nesting.

File: client/js/lounge.js (hunk 3 above)
Comment (YaManicKill) [reply client/js/lounge.js]: Sorry, I never did elaborate, but it looks like you figured out what I meant.

File: client/js/utils.js [hunk 9]
```diff
@@ -29,6 +43,27 @@ function forceFocus() {
 	input.trigger("click").focus();
 }
 
+function collapse() {
```
Comment (YaManicKill) [inline client/js/utils.js]: (I know I said the opposite of this previously, so this is not for this PR...it's for the future)

If we get more than this, we will probably want to split them up out of `utils.js`, but not into just another file. Probably have a directory with a file per command. But as I say, not for this PR.
